From 123f34de4206d38574e36552776c1414e577b4bf Mon Sep 17 00:00:00 2001
From: Oliver Tappe <zooey@hirschkaefer.de>
Date: Fri, 6 Jan 2017 11:47:08 +0100
Subject: [PATCH] gupnp-connman-manager: Handle IP-address changes gracefully.

* React to IP-address changes by recreating the UPNP-context, such that
  the new IP-address will be picked up properly.
---
 libgupnp/gupnp-connman-manager.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/libgupnp/gupnp-connman-manager.c b/libgupnp/gupnp-connman-manager.c
index 6a48beb..fc7714b 100644
--- a/libgupnp/gupnp-connman-manager.c
+++ b/libgupnp/gupnp-connman-manager.c
@@ -227,7 +227,7 @@ on_service_property_changed (CMService   *cm_service,
                     g_strcmp0 (cm_service->iface,
                                gssdp_client_get_interface (GSSDP_CLIENT (cm_service->context))) != 0) {
                         service_context_delete (cm_service);
-                        service_context_create (cm_service);
+                        service_context_install_creation_timeout (cm_service);
                 }
 
         } else if (g_strcmp0 (name, "State") == 0) {
@@ -240,6 +240,18 @@ on_service_property_changed (CMService   *cm_service,
                         new_state = CM_SERVICE_STATE_INACTIVE;
 
                 service_context_update (cm_service, new_state);
+
+        } else if (g_strcmp0 (name, "IPv4") == 0) {
+                const char *newAddress = NULL;
+                g_variant_lookup (value, "Address", "&s", &newAddress);
+
+                if (cm_service->context != NULL &&
+                    g_strcmp0 (newAddress,
+                               gssdp_client_get_host_ip (GSSDP_CLIENT (cm_service->context))) != 0) {
+                        service_context_delete (cm_service);
+                        service_context_install_creation_timeout (cm_service);
+                }
+
         }
 }
 
-- 
2.7.4

